cargo-quickbuild/src/main.rs

```diff
@@ -5,59 +5,89 @@ use cargo_lock::{
+        deps.insert(graph[n].clone());
+        let sub_neighbours = get_dependencies(graph, &n);
+        deps.extend(sub_neighbours);
+    }
+    deps
+}
+
+fn hash_packages(packages: &BTreeSet<Package>) -> String {
+    let mut hasher = Sha256::new();
+    let debugged = format!("{:?}", packages);
+    hasher.update(debugged);
+    format!("{:x}", hasher.finalize())
+}
+
 fn main() -> Result<(), Error> {
     let lockfile = Lockfile::load("Cargo.lock").unwrap();
     // FIXME: if lockfile.metadata or lockfile.patch contain anything
     // interesting then explode.
     let tree = lockfile.dependency_tree()?;
     let graph = tree.graph();
 
-    for index in tree.roots().iter() {
-        let lockfiles = make_lockfiles(&lockfile, &graph, *index);
-        for lockfile in lockfiles {
-            // FIXME: find root crate name here
-            let root = "subtree";
-            let contents = lockfile.to_string();
-            let hash = hash(&contents);
-            let filename = format!("{root}-{hash}.lock", root = root, hash = hash);
-            std::fs::write(filename, contents).expect("Unable to write file");
-        }
+    for node in tree.nodes().iter() {
+        let (dependency, node_index) = node;
+        let deps = get_dependencies(&graph, &node_index);
+        let hash = hash_packages(&deps);
+
+        println!("{}-{}", dependency.name.as_str(), hash);
     }
+
     Ok(())
 }
 
-fn make_lockfiles(lockfile: &Lockfile, graph: &Graph, index: NodeIndex) -> Vec<Lockfile> {
-    let mut result = vec![make_subtree_lockfile(lockfile, graph, index)];
+#[cfg(test)]
+mod test {
+    use std::path::PathBuf;
 
-    for ix in graph.neighbors(index) {
-        result.extend(make_lockfiles(lockfile, graph, ix).into_iter());
-    }
-    result
-}
+    use super::*;
 
-fn make_subtree_lockfile(lockfile: &Lockfile, graph: &Graph, index: NodeIndex) -> Lockfile {
-    let packages = walk_subtree(graph, index);
-    let mut set = BTreeSet::new();
-    set.extend(packages.into_iter());
-    let packages = set.into_iter().collect();
+    #[test]
+    fn test_get_dependencies() {
+        let mut d = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        d.push("../Cargo.lock");
+        let lockfile = Lockfile::load(d).unwrap();
+        let tree = lockfile.dependency_tree().unwrap();
+        let graph = tree.graph();
+        let (_dep, node_index) = tree
+            .nodes()
+            .iter()
+            .find(|(dep, _node_index)| dep.name.as_str() == "serde")
+            .unwrap();
 
-    Lockfile {
-        packages,
-        ..lockfile.clone()
-    }
-}
+        let packages = get_dependencies(&graph, &node_index);
 
-fn walk_subtree(graph: &Graph, index: NodeIndex) -> Vec<Package> {
-    let mut result = vec![graph[index].clone()];
-    for ix in graph.neighbors(index) {
-        result.extend(walk_subtree(graph, ix).into_iter());
+        let package_names = vec!["proc-macro2", "quote", "serde_derive", "syn", "unicode-xid"];
```

Comment:
I just did `cargo run` and it spat out the 44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a for all of the 0-dependency packages ( `cfg-if` and `version_check` and `hashbrown` and `matches` are crates that have 0 dependencies and they end up with the same hash). I think we want to include `serde` in the result from `get_dependencies(&graph, &node_index_for_serde)` (maybe call it `get_dependencies_including_self(&graph, &node_index)` or something?). This way, when someone says "give me the hash of the dependency tree from serde downwards", it also includes the version of serde, and we won't get confused if they change the version of serde they're using.

I can submit a patch for this if you want.